internal/vhdx/info.go

```diff
@@ -0,0 +1,233 @@
+	PartitionStyle uint32
+	PartitionCount uint32
+	// A union of driveLayoutInformationGPT and driveLayoutInformationMBR
+	// since driveLayoutInformationGPT is largest with 40 bytes
+	GptMbrUnion    [40]byte
+	PartitionEntry [1]partitionInformationEx
+}
+
+// Takes the handle to an attached vhdx and retrieves the drive layout information of that vhdx.
+// Returns the driveLayoutInformationEx struct and a slice of partitionInfomrationEx struct containing
+// one element for each partition found on the vhdx. Note: some of the members like (GptMbrUnion) of these
+// structs are raw byte array and it is the responsibility of the calling function to properly parse them.
+func getDriveLayout(ctx context.Context, diskHandle syscall.Handle) (driveLayoutInformationEx, []partitionInformationEx, error) {
+	var (
+		outBytes   uint32
+		err        error
+		volume     *os.File
+		volumePath string
+	)
+
+	layoutData := struct {
+		info driveLayoutInformationEx
+		// driveLayoutInformationEx has a flexible array member at the end. The data returned
+		// by IOCTL_DISK_GET_DRIVE_LAYOUT_EX usually has driveLayoutInformationEx.PartitionCount
+		// number of elements in this array. For all practical purposes we don't expect to have
+		// more than 64 partitions in a container/uvm vhdx.
+		partitions [63]partitionInformationEx
+	}{}
+
+	volumePath, err = vhd.GetVirtualDiskPhysicalPath(diskHandle)
+	if err != nil {
+		return layoutData.info, layoutData.partitions[:0], err
+	}
+
+	volume, err = os.OpenFile(volumePath, os.O_RDONLY, 0)
+	if err != nil {
+		return layoutData.info, layoutData.partitions[:0], fmt.Errorf("failed to open drive: %s", err)
+	}
+	defer volume.Close()
+
+	err = windows.DeviceIoControl(windows.Handle(volume.Fd()),
+		_IOCTL_DISK_GET_DRIVE_LAYOUT_EX,
+		nil,
+		0,
+		(*byte)(unsafe.Pointer(&layoutData)),
+		uint32(unsafe.Sizeof(layoutData)),
+		&outBytes,
+		nil)
+	if err != nil {
+		return layoutData.info, layoutData.partitions[:0], fmt.Errorf("IOCTL to get disk layout failed: %s", err)
+	}
+
+	if layoutData.info.PartitionCount == 0 {
+		return layoutData.info, []partitionInformationEx{}, nil
+	} else {
+		// parse the retrieved data into driveLayoutInformationEx and partitionInformationEx
+		partitions := make([]partitionInformationEx, layoutData.info.PartitionCount)
+		partitions[0] = layoutData.info.PartitionEntry[0]
+		copy(partitions[1:], layoutData.partitions[:layoutData.info.PartitionCount-1])
+		return layoutData.info, partitions, nil
+	}
+}
+
+// Scratch VHDs are formatted with GPT style and have 1 MSFT_RESERVED
+// partition and 1 BASIC_DATA partition.  This struct contains the
+// partitionID of this BASIC_DATA partition and the DiskID of this
+// scratch vhdx.
+type ScratchVhdxPartitionInfo struct {
+	DiskID      guid.GUID
+	PartitionID guid.GUID
+}
+
+// Returns the VhdxInfo of a GPT vhdx at path vhdxPath.
+func GetScratchVhdPartitionInfo(ctx context.Context, vhdxPath string) (_ ScratchVhdxPartitionInfo, err error) {
```

Comment:
We can move it inside the `computestorage` package where this will be used but then we will have to move most of the type definitions (like `driveLayout`, `partitionInformationEx` etc.) from this package over there as well. I think keeping those vhd specific types in this package makes more sense. Moreover, since this package is internal, it also conveys that this is vhdx related stuff specific to hcsshim and not for generic use. 
What do you think?

Reply:
That seems fine to me. We can keep it here for now and always move it later if we change our minds.